File: IPWorks Samples/XMPP/cpp/jabber.cpp
/*
 * IPWorks 2022 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworks
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */


#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "../../include/ipworks.h"
#define LINE_LEN 100

class MyXMPP : public XMPP
{
public:
	virtual int FireConnected(XMPPConnectedEventParams *e)
	{
		printf("Welcome! Connection established!\n");
		return 0;
	}
	virtual int FireDisconnected(XMPPDisconnectedEventParams *e)
	{
		printf("Disconnected\n");
		return 0;
	}
	virtual int FireMessageIn(XMPPMessageInEventParams *e)
	{
		printf("%s said : %s\n\n", e->From, e->MessageText);
		return 0;
	}

	virtual int FireSSLServerAuthentication(XMPPSSLServerAuthenticationEventParams *e)
	{
		e->Accept = true;
		return 0;
	}
};

int main(int argc, char* argv[])
{

	if (argc != 4) {

		fprintf(stderr, "usage: jabber server username password\n\n");
		fprintf(stderr, "  server    the name or address of the jabber server (XMPP server)\n");
		fprintf(stderr, "  username  the user name used to authenticate to the jaber server\n");
		fprintf(stderr, "  password  the password used to authenticate to the jaber server\n");
		fprintf(stderr, "\nExample: jabber talk.google.com username password\n\n");
		printf("Press enter to continue.");
		getchar();
	}
	else{

		MyXMPP jabber;
		char buddy[80]; //users input
		char buffer[LINE_LEN + 1];
		int ret_code = 0;
		int buddyCount = 0;


		jabber.SetIMServer(argv[1]);
		jabber.SetSSLStartMode(SSL_AUTOMATIC);
		
		printf("Connecting...\n\n");
		ret_code = jabber.ConnectTo(argv[2], argv[3]);
		if (ret_code) goto done;

		buddyCount = jabber.GetBuddyCount();
		printf("Buddy list:\n");

		if (buddyCount != 0){

			for (int i = 0; i < buddyCount; i++)
			{
				printf("%d""%s""%s\n", i + 1, ") ", jabber.GetBuddyId(i));
			}

			printf("\nSelect a buddy : ");
			fgets(buddy, 80, stdin);
			buddy[strlen(buddy) - 1] = '\0';

			printf("\nMessage: ");
			fgets(buffer, 80, stdin);
			buffer[strlen(buffer) - 1] = '\0';

			jabber.SetMessageText(buffer);

			jabber.SendMessage(jabber.GetBuddyId(atoi(buddy) - 1));

			printf("\n\nReceiving responses, Ctrl-C to exit.\n");
			while (1)
			{
				jabber.DoEvents();
			}

			printf("Disconnecting...\n\n");
			ret_code = jabber.Disconnect();
			if (ret_code) goto done;

		}
		else
		{
			printf("No buddies found!");
		}

	done:

		if (ret_code)     // Got an error.  The user is done.
		{
			printf("\nError: %d", ret_code);
			if (jabber.GetLastError())
			{
				printf(" \"%s\"\n", jabber.GetLastError());
			}
		}
		fprintf(stderr, "\npress <return> to continue...\n");
		getchar();
		exit(ret_code);
		return 0;


	}

}

File: IPWorks Samples/XMPP/javascript/jabber.js
/*
 * IPWorks 2024 JavaScript Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworks
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */
 
const readline = require("readline");
const ipworks = require("@nsoftware/ipworks");

if(!ipworks) {
  console.error("Cannot find ipworks.");
  process.exit(1);
}
let rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

main();

async function main() {
  const argv = process.argv;
  if (argv.length !== 5) {
    console.log("Usage: node jabber.js server user password");
    console.log("  server    the address of the XMMP (Jabber) server");
    console.log("  user      the user associated with the server");
    console.log("  password  the password to logon");
    console.log("Example: node jabber.js servername username@gmail.com password");
    process.exit();
  }

  var messageReceived = false;
  const xmpp = new ipworks.xmpp();

  xmpp.on("SSLServerAuthentication", function (e) {
    e.accept = true;
  })
	.on("MessageIn", function (e) {
    console.log(e.from + " said : " + e.messageText);
    messageReceived = true;
  });

  try {
    xmpp.setIMServer(argv[2]);
    xmpp.setIMPort(5222);
    xmpp.setSSLEnabled(true);
    xmpp.setSSLStartMode(2);
    xmpp.config("AcceptAnyServerCert=true"); // In most cases certificate will be unauthorized, so we have to add this
    console.log("Connecting...");
    console.error(argv[2], argv[3], argv[4]);
    await xmpp.connectTo(argv[3], argv[4]);
    console.log("Welcome! Connection established!");
  } catch (e) {
    console.log(e);
    process.exit();
  }

  let buddyCount = xmpp.getBuddies().size();
  let buddy = 0;
  console.log("Buddy list:");

  if (parseInt(buddyCount) !== 0) {
    for (i = 0; i < buddyCount; i++) {
      console.log(i + 1 + ") " + xmpp.getBuddies().item(i).getId());
    }

    rl.question("Select a buddy:", (answer) => {
      buddy = parseInt(answer);
      rl.question("Message to send to buddy:", async (answer) => {
        xmpp.setMessageText(answer);
        console.log("Sending...");
        await xmpp.sendMessage(xmpp.getBuddies().item(buddy - 1).getId());
        rl.question("Enter 'y' to wait for a response [n]: ", async (answer) => {
          if (answer.toLowerCase().startsWith("y")) {
            console.log("Waiting for response...");
            while (messageReceived === false) {
              await xmpp.doEvents();
            }
          }
          console.log("Disconnecting...");
          xmpp.disconnect();
          console.log("Bye!");
          rl.close();
        });
      });
    });
  } else {
    console.log("No buddies found!");
  }
}

function prompt(promptName, label, punctuation, defaultVal)
{
  lastPrompt = promptName;
  lastDefault = defaultVal;
  process.stdout.write(`${label} [${defaultVal}] ${punctuation} `);
}
